cdap-app-fabric/src/main/java/co/cask/cdap/runtime/monitor/RuntimeMonitor.java

```diff
@@ -0,0 +1,177 @@
+import co.cask.cdap.client.config.ClientConfig;
+import co.cask.cdap.client.util.RESTClient;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.logging.LogSampler;
+import co.cask.cdap.common.logging.LogSamplers;
+import co.cask.cdap.common.logging.Loggers;
+import co.cask.cdap.proto.id.NamespaceId;
+import co.cask.common.http.HttpMethod;
+import co.cask.common.http.HttpResponse;
+import com.google.common.base.Supplier;
+import com.google.common.reflect.TypeToken;
+import com.google.common.util.concurrent.AbstractExecutionThreadService;
+import com.google.gson.Gson;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Type;
+import java.net.HttpURLConnection;
+import java.nio.charset.Charset;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.ws.rs.ServiceUnavailableException;
+
+/**
+ * Runtime Monitor Runnable responsible for fetching metadata
+ */
+public class RuntimeMonitor extends AbstractExecutionThreadService {
+  private static final Logger LOG = LoggerFactory.getLogger(RuntimeMonitor.class);
+  // For outage, only log once per 60 seconds per message.
+  private static final Logger OUTAGE_LOG = Loggers.sampling(LOG, LogSamplers.perMessage(new Supplier<LogSampler>() {
+    @Override
+    public LogSampler get() {
+      return LogSamplers.limitRate(60000);
+    }
+  }));
+
+  private static final Gson GSON = new Gson();
+  private static final Type MAP_STRING_STRING_TYPE = new TypeToken<Map<String, String>>() { }.getType();
+  private static final Type MAP_STRING_MESSAGE_TYPE = new TypeToken<Map<String, List<MonitorMessage>>>() { }.getType();
+
+  private final RESTClient restClient;
+  private final ClientConfig clientConfig;
+
+  private final int limit;
+  private final String runtimeId;
+  private final CConfiguration cConf;
+  private Map<String, MonitorConsumeRequest> topicsToRequest;
+
+  private final MessagePublisher messagePublisher;
+  private final long pollTimeMillis;
+  private volatile Thread runThread;
+  private volatile boolean stopped;
+
+  public RuntimeMonitor(String runtimeId, CConfiguration cConf, MessagePublisher messagePublisher,
+                        ClientConfig clientConfig) {
+    this.runtimeId = runtimeId;
+    this.cConf = cConf;
+    this.messagePublisher = messagePublisher;
+    this.clientConfig = clientConfig;
+    this.restClient = new RESTClient(clientConfig);
+    this.limit = cConf.getInt(Constants.RuntimeMonitor.RUNTIME_MONITOR_BATCH_LIMIT);
+    this.pollTimeMillis = cConf.getInt(Constants.RuntimeMonitor.RUNTIME_MONITOR_POLLTIME_MS);
+    this.topicsToRequest = new HashMap<>();
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.debug("Starting runtime monitor {}....", runtimeId);
+    initializeTopics();
```

Comment:
This method shouldn't fail as well. Better make the initialization on-demand and only make the REST call if it is not yet initialized. You can do it as part of the monitor loop so that you can handle exception in the same way.